src/Microsoft.Azure.WebJobs.Host/Bindings/Path/BindingTemplateToken.cs

```diff
@@ -1,32 +1,273 @@
-        { 
-            get { return _value; } 
+        // Represent an expression. 
+        //    "literal{expr}literal" 
+        private class ExpressionToken : BindingTemplateToken
+        {
+            private readonly string[] _parts;
+
+            public ExpressionToken(string[] parts)
+            {
+                _parts = parts;
+            }
+
+            public override string AsLiteral
+            {
+                get
+                {
+                    return null;
+                }
+            }
+
+            public override string ParameterName
+            {
+                get
+                {
+                    return this._parts[0];
+                }
+            }
+
+            protected override void BuildCapturePattern(StringBuilder builder)
+            {
+                if (this._parts.Length != 1)
+                {
+                    throw new InvalidOperationException($"Capture expressions can't include dot operators");
+                }
+
+                var value = this._parts[0];
+                builder.Append(String.Format(CultureInfo.InvariantCulture, "(?<{0}>.*)", value));
+            }
+
+            protected virtual string Fallback()
+            {
+                // Derived class can provide fallback resolutions if we don't match binding data. 
+                return null;
+            }
+
+            public override string Evaluate(IReadOnlyDictionary<string, object> bindingData)
+            {
+                object current;
+
+                if (bindingData.TryGetValue(this._parts[0], out current))
+                {
+                    for (int i = 1; i < _parts.Length; i++)
+                    {
+                        var propName = _parts[i];
+
+                        try
+                        {
+                            current = GetProperty(current, propName);
+                        }
+                        catch (Exception e)
+                        {
+                            throw new InvalidOperationException($"Error while accessing '{propName}': {e.Message}");
+                        }
+                    }
+                }
+                else
+                {
+                    // Not found. Maybe a resolver can do it?
+                    current = this.Fallback();
+
+                    if (current == null)
+                    {
+                        throw new InvalidOperationException(String.Format(CultureInfo.CurrentCulture, "No value for named parameter '{0}'.", this.ParameterName));
+                    }
+                }
+
+                var strValue = BindingDataPathHelper.ConvertParameterValueToString(current);
+                return strValue;
+            }
+
+            // Case insensitive lookup.
+            // Helper to get a property from an object.             
+            // This supports both static types and JObject (for dynamic) 
+            public static object GetProperty(object o, string member)
```

Comment:
This seems like a lot of reflection to run for each execution. Look at the `PropertyHelper` class - this would likely be more performant if we cached an expression tree of fast property getters to do the dot expression evaluation